File: leetcode/0755-pour-water/Solution.java
class Solution {
    public int[] pourWater(int[] heights, int volume, int k) {
        int n = heights.length;

        for (int drop = 0; drop < volume; drop++) {
            int best = k;

            int i = k;
            while (i - 1 >= 0 && heights[i - 1] <= heights[i]) {
                i--;
                if (heights[i] < heights[best]) {
                    best = i;
                }
            }
            if (best != k) {
                heights[best]++;
                continue;
            }

            i = k;
            while (i + 1 < n && heights[i + 1] <= heights[i]) {
                i++;
                if (heights[i] < heights[best]) {
                    best = i;
                }
            }

            heights[best]++;
        }

        return heights;
    }
}

File: leetcode/0755-pour-water/solution.py
from typing import List


class Solution:
    def pourWater(self, heights: List[int], volume: int, k: int) -> List[int]:
        n = len(heights)

        for _ in range(volume):
            best = k

            # try to flow left
            i = k
            while i - 1 >= 0 and heights[i - 1] <= heights[i]:
                i -= 1
                if heights[i] < heights[best]:
                    best = i
            if best != k:
                heights[best] += 1
                continue

            # try to flow right
            i = k
            while i + 1 < n and heights[i + 1] <= heights[i]:
                i += 1
                if heights[i] < heights[best]:
                    best = i

            heights[best] += 1

        return heights
